homeassistant/components/forecast_solar/__init__.py

```diff
@@ -1,66 +1,19 @@
 """The Forecast.Solar integration."""
 from __future__ import annotations
 
-from datetime import timedelta
-import logging
-
-from forecast_solar import Estimate, ForecastSolar
-
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE, Platform
+from homeassistant.const import Platform
 from homeassistant.core import HomeAssistant
-from homeassistant.helpers.aiohttp_client import async_get_clientsession
-from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
-from .const import (
-    CONF_AZIMUTH,
-    CONF_DAMPING,
-    CONF_DECLINATION,
-    CONF_INVERTER_SIZE,
-    CONF_MODULES_POWER,
-    DOMAIN,
-)
+from .const import DOMAIN
+from .coordinator import ForecastSolarDataUpdateCoordinator
 
 PLATFORMS = [Platform.SENSOR]
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Forecast.Solar from a config entry."""
-    # Our option flow may cause it to be an empty string,
-    # this if statement is here to catch that.
-    api_key = entry.options.get(CONF_API_KEY) or None
-
-    if (
-        inverter_size := entry.options.get(CONF_INVERTER_SIZE)
-    ) is not None and inverter_size > 0:
-        inverter_size = inverter_size / 1000
-
-    session = async_get_clientsession(hass)
-    forecast = ForecastSolar(
-        api_key=api_key,
-        session=session,
-        latitude=entry.data[CONF_LATITUDE],
-        longitude=entry.data[CONF_LONGITUDE],
-        declination=entry.options[CONF_DECLINATION],
-        azimuth=(entry.options[CONF_AZIMUTH] - 180),
-        kwp=(entry.options[CONF_MODULES_POWER] / 1000),
-        damping=entry.options.get(CONF_DAMPING, 0),
-        inverter=inverter_size,
-    )
-
-    # Free account have a resolution of 1 hour, using that as the default
-    # update interval. Using a higher value for accounts with an API key.
-    update_interval = timedelta(hours=1)
-    if api_key is not None:
-        update_interval = timedelta(minutes=30)
-
-    coordinator: DataUpdateCoordinator[Estimate] = DataUpdateCoordinator(
-        hass,
-        logging.getLogger(__name__),
-        name=DOMAIN,
-        update_method=forecast.estimate,
-        update_interval=update_interval,
-    )
+    coordinator = ForecastSolarDataUpdateCoordinator(hass, entry)
```

Comment:
I would have preferred injecting the forecast object into the coordinator. It makes testing easier and avoids config structure dependencies in the coordinator.

Reply:
I personally do not, this makes the coordinator responsible for any data and connection handling.

There is no difference in testing, as this PR shows (besides a small patch location change).